### PR TITLE
gh-95382: Use cache for indentations in the JSON encoder

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
It is a continuation of #118105 which keeps a cache of newline+indentation and comma+newline+indentation at every level.

For example:
```
$ ./python -m timeit -s 'import json; a = [[[{"key": "value"}]*10]*10]*10' 'json.dumps(a, indent=2)'
Before: 200 loops, best of 5: 1.79 msec per loop
After:  200 loops, best of 5: 1.25 msec per loop
```
```
$ ./python -m timeit -s 'import json; a = [[[list(range(10))]*10]*10]*10' 'json.dumps(a, indent=2)'
Before: 50 loops, best of 5: 5.45 msec per loop
After:  50 loops, best of 5: 4.78 msec per loop
```
The effect is the strongest if there are many deep narrow trees growing from the root:
```
$ ./python -m timeit -s 'import json; a = [[[["nested"]]]]*1000' 'json.dumps(a, indent=2)'
100 loops, best of 5: 3.39 msec per loop
200 loops, best of 5: 1.85 msec per loop
```
And the weakest (up to no difference) if the tree contains few large flat lists or dicts.

* Issue: gh-95382

### CODE REVIEW

File: Modules/_json.c [hunk 1]
```diff
@@ -1720,11 +1763,8 @@ encoder_listencode_list(PyEncoderObject *s, _PyUnicodeWriter *writer,
     }
 
     if (s->indent != Py_None) {
-        Py_CLEAR(new_newline_indent);
-        Py_CLEAR(separator_indent);
-        if (_PyUnicodeWriter_WriteStr(writer, newline_indent) < 0) {
-            goto bail;
-        }
+        indent_level--;
+        do_dedent(s, writer, indent_level, indent_cache);
```
Comment (eendebakpt) [inline Modules/_json.c]: The result value of `do_dedent` should be handled right?
```suggestion
        if (do_dedent(s, writer, indent_level, indent_cache) < 0) {
            goto bail;
        }
```

File: Modules/_json.c [hunk 2]
```diff
@@ -1252,14 +1252,81 @@ encoder_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
 }
 
 static PyObject *
-_create_newline_indent(PyObject *indent, Py_ssize_t indent_level)
+_create_indent_cache(PyEncoderObject *s, Py_ssize_t indent_level)
```
Comment (eendebakpt) [inline Modules/_json.c]: For consistency I would remove the leading underscore (or add a leading underscore to `update_newline_indent`, `do_indent`, ...)

File: Modules/_json.c [hunk 3]
```diff
@@ -1252,14 +1252,81 @@ encoder_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
...
     if (newline_indent != NULL && indent_level) {
         PyUnicode_AppendAndDel(&newline_indent,
-                               PySequence_Repeat(indent, indent_level));
+                               PySequence_Repeat(s->indent, indent_level));
     }
-    return newline_indent;
+    if (newline_indent == NULL) {
+        return NULL;
+    }
+    PyObject *indent_cache = PyList_New(2);
```
Comment (eendebakpt) [inline Modules/_json.c]: Only a single cache is used for both the newline indent and seperator indent (at the even and odd indices of the list). A comment or docstring in this method would help to clarify.

File: Modules/_json.c [hunk 4]
```diff
@@ -1629,12 +1688,8 @@ encoder_listencode_dict(PyEncoderObject *s, _PyUnicodeWriter *writer,
...
     }
     if (s->indent != Py_None) {
-        Py_CLEAR(new_newline_indent);
-        Py_CLEAR(separator_indent);
-
-        if (_PyUnicodeWriter_WriteStr(writer, newline_indent) < 0) {
-            goto bail;
-        }
+        indent_level--;
+        do_dedent(s, writer, indent_level, indent_cache);
```
Comment (serhiy-storchaka) [inline Modules/_json.c]:
```suggestion
        if (do_dedent(s, writer, indent_level, indent_cache) < 0) {
            goto bail;
        }
```

File: Modules/_json.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/_json.c]: Right.

File: Modules/_json.c [hunk 5]
```diff
@@ -1278,20 +1353,20 @@ encoder_call(PyEncoderObject *self, PyObject *args, PyObject *kwds)
     _PyUnicodeWriter_Init(&writer);
     writer.overallocate = 1;
 
-    PyObject *newline_indent = NULL;
+    PyObject *indent_cache = NULL;
     if (self->indent != Py_None) {
-        newline_indent = _create_newline_indent(self->indent, indent_level);
-        if (newline_indent == NULL) {
+        indent_cache = create_indent_cache(self, indent_level);
```
Comment (eendebakpt) [inline Modules/_json.c:1360]: When called via the public interface `indent_level` is always zero. By removing the argument `indent_level` in the internal implementation this call and some other bits of the code can be a bit simplified. 

(this is probably not worth creating a separate PR for, I leave it up to you)

File: Modules/_json.c [hunk 6]
```diff
@@ -1251,17 +1251,92 @@ encoder_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
...
+    }
+    assert(indent_level * 2 < PyList_GET_SIZE(indent_cache));
+    return PyList_GET_ITEM(indent_cache, indent_level * 2 - 1);
+}
+
+static int
+write_newline_indent(_PyUnicodeWriter *writer,
+                     Py_ssize_t indent_level, PyObject *indent_cache)
+{
+    PyObject *newline_indent = PyList_GET_ITEM(indent_cache, indent_level * 2);
```
Comment (eendebakpt) [inline Modules/_json.c:1336]:
```suggestion
    assert(indent_level * 2 < PyList_GET_SIZE(indent_cache));
    PyObject *newline_indent = PyList_GET_ITEM(indent_cache, indent_level * 2);
```

File: Modules/_json.c [hunk 7]
```diff
@@ -1251,17 +1251,92 @@ encoder_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
...
+get_item_separator(PyEncoderObject *s,
+                   Py_ssize_t indent_level, PyObject *indent_cache)
+{
+    assert(indent_level > 0);
+    if (indent_level * 2 > PyList_GET_SIZE(indent_cache)) {
+        if (update_indent_cache(s, indent_level, indent_cache) < 0) {
+            return NULL;
+        }
+    }
+    assert(indent_level * 2 < PyList_GET_SIZE(indent_cache));
```
Comment (eendebakpt) [inline Modules/_json.c:1328]:
```suggestion
    assert(indent_level * 2 - 1 < PyList_GET_SIZE(indent_cache));
```

**Review by eendebakpt — APPROVED** [review]
The PR adds some extra code for creation and usage of the cache, but this simplifies the error handling in the methods using the cache. The performance improvement is a bit larger than I expected which is nice, although it is also fair to say that the performance when writing json files with `indent != None` is probably not the most important.

File: Modules/_json.c (hunk 7 above)
Comment (serhiy-storchaka) [reply Modules/_json.c:1328]: `indent_level * 2 < PyList_GET_SIZE(indent_cache)` is more strong condition.

File: Modules/_json.c (hunk 6 above)
Comment (serhiy-storchaka) [reply Modules/_json.c:1336]: The assert above is placed after conditional resizing the cache. It is useful to check that we always get the right cache size, independently from conditions. But here we do not change the cache.

**Review by eendebakpt — APPROVED** [review]

**Pull request merged**